src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs

```diff
@@ -91,15 +89,65 @@ static string DecodeFrom64(string encodedData)
                 var operation = TableOperation.Delete(subscription);
                 await table.ExecuteAsync(operation).ConfigureAwait(false);
             }
+            ClearForMessageType(messageType);
+        }
+
+        static string GetKey(List<MessageType> types)
+        {
+            var typeNames = types.Select(_ => _.TypeName);
+            return string.Join(",", typeNames) + ",";
+        }
+
+        static string GetKeyPart(MessageType type)
+        {
+            return $"{type.TypeName},";
+        }
+
+        void ClearForMessageType(MessageType messageType)
+        {
+            if (cacheFor == null)
+            {
+                return;
+            }
+            var keyPart = GetKeyPart(messageType);
+            foreach (var cacheKey in Cache.Keys)
+            {
+                if (cacheKey.Contains(keyPart))
+                {
+                    CacheItem cacheItem;
+                    Cache.TryRemove(cacheKey, out cacheItem);
+                }
+            }
         }
 
-        /// <summary>
-        /// Returns the subscription address based on message type
-        /// </summary>
-        /// <param name="messageTypes">Types of messages that subscription addresses should be found for</param>
-        /// <param name="context">The current pipeline context</param>
-        /// <returns>Subscription addresses that were found for the provided messageTypes</returns>
-        public async Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes, ContextBag context)
+        public Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes, ContextBag context)
+        {
+            var types = messageTypes.ToList();
+
+            if (cacheFor == null)
+            {
+                return GetSubscriptions(types);
+            }
+
+            var key = GetKey(types);
+
+            var cacheItem = Cache.GetOrAdd(key,
+                valueFactory: _ => new CacheItem
+                {
+                    Stored = DateTime.UtcNow,
+                    Subscribers = GetSubscriptions(types)
+                });
+
+            var age = DateTime.UtcNow - cacheItem.Stored;
+            if (age >= cacheFor)
+            {
+                cacheItem.Subscribers = GetSubscriptions(types);
```

Comment:
I agree with @Scooletz using the update APIs of the concurrent dictionary might be a more clear approach